vulkan_generator/vulkan_parser/struct_parser.py

```diff
@@ -0,0 +1,153 @@
+    """Parses a Vulkan Struct member
+
+    This is a bit of an irregular code because the XML itself has quite irregularities that
+    makes is hard to parse type and variable easily.
+
+    For example a const pointer member is defined as:
+     <member optional="true">const <type>void</type>*     <name>pNext</name></member>
+
+    Where as a static array defined as:
+    <member limittype="noauto">
+        <type>char</type>
+        <name>deviceName</name>[<enum>VK_MAX_PHYSICAL_DEVICE_NAME_SIZE</enum>]
+    </member>
+    """
+
+    #  This is a bit of an irregular code because the XML itself has quite irregularities that
+    # makes is hard to parse type and variable easily.
+    #
+    # This is not the code we wanted but it's the code that we needed and it's contained in a
+    # small place so that XML irregularities does not leak into the rest of the code.
+
+    members = {}
+    for member_element in struct_element:
+        if member_element.tag == "comment":
+            # Melih TODO: We may want to support comments in the future
+            continue
+
+        if member_element.tag != "member":
+            raise SyntaxError(f"No member tag found in : {ET.tostring(member_element, 'utf-8')}")
+
+        typename = parsing_utils.get_text_from_tag(member_element, "type")
+        variable_name = parsing_utils.get_text_from_tag(member_element, "name")
+
+        # Type attributes(const, struct) and pointer attributes(*, const*, *const,*const*)
+        # are usually in the text field of the member tag.
+        #
+        # Below is the example for a const pointer:
+        #
+        # <member optional="true">const <type>void</type>*     <name>pNext</name></member>
+        #
+        # In the below type is "const char* const*" but only "char" is in the type tag.
+        # The rest of the information is scattered around the member's text
+        # Therefore we need to retrieve and clean it so we can add it to the type.
+        #
+        # <member len="enabledLayerCount,null-terminated">const <type>char</type>
+        # * const*      <name>ppEnabledLayerNames</name>
+        #
+        type_attributes = parsing_utils.try_get_text_from_tag(member_element, "member")
+        # some times it's just empty space or endline character
+        if type_attributes is not None:
+            type_attributes = parsing_utils.clean_type_string(type_attributes)
+
+        if type_attributes and len(type_attributes) > 0:
+            typename = f"{type_attributes} {typename}"
+
+        pointers = parsing_utils.try_get_tail_from_tag(member_element, "type")
+        if pointers is not None:
+            pointers = parsing_utils.clean_type_string(pointers)
+            # Add space between "*" and "const"
+            pointers = pointers.replace("const", " const")
+            typename = f"{typename}{pointers}"
+
+        if typename is None:
+            raise SyntaxError(f"No typename found in : {ET.tostring(member_element, 'utf-8')}")
```

Comment:
That's why I wanted to raise exception. I don't know if it's pythonic(please let me know the correct way if it's not) but that's what I was trying to tell here.  If the typename is `None` here it's wrong. Maybe I should check both `None` and empty string. Would it be:

`if not typename:` ? Or something similar.

Basically general principle is, fail as early as possible if something is wrong. Because we know the XML that we are parsing, it shouldn't go wrong and if we update the XML and things changes, then we would know what has changed.

Reply:
Oh this is fine. I guess my point was more: `typename` cannot be `None` according to the type hints. So, if you ever violated this in the future, `mypy` should fail. So, you could rely on the type hints to catch this, instead of using an exception or assert. It would be similar to if the type of `typename` was `std::string` in C++; it cannot be nullptr, so you would not check for this (although you might check that isn't zero length).

Even if you don't want to rely on the type hints, some people would argue that you should use:

```python
assert typename, "Expected typename to have a non-zero value!"
```

here instead of an exception, as `typename` should always be set, regardless of input values (the XML, I guess). I.e. it is guaranteed by the (lack of) control flow above, and by the fact that `parsing_utils.get_text_from_tag` cannot return None, and is not affected by the XML input. Actually, perhaps this is not true because the XML could contain a zero-length string. Regardless, I would be fine with it. On the other hand, some people say that you should not use `assert` outside tests. The Google style seems to allow/encourage asserts for checking "internal correctness", which is what I mean here.
 
If you still want to use an `if`, then yes, I suggest:

```
if not typename:
  raise ...
```

Reply:
👍